cou/steps/analyze.py

```diff
@@ -18,20 +18,44 @@
 
 import logging
 import os
-from dataclasses import dataclass
-from typing import Iterable
+from collections import defaultdict
+from dataclasses import dataclass, field
+from typing import Iterable, List, Optional
 
 from cou.apps.app import Application
 from cou.utils.juju_utils import async_get_application_config, async_get_status
+from cou.utils.openstack import (
+    LTS_SERIES,
+    SPECIAL_CHARMS,
+    UPGRADE_ORDER,
+    OpenStackRelease,
+)
 
 logger = logging.getLogger(__name__)
 
 
 @dataclass
 class Analysis:
-    """Analyze result."""
+    """Analyze result.
+
+    :param apps: Applications in the model
+    :type apps:  Iterable[Application]
+    :param apps_to_upgrade: Applications to upgrade in the model
```

Comment:
Why are we determining a list of applications to upgrade here? I thought we had agreed in the spec that it would be simpler to have a predetermined upgrade process to iterate through, and have each Application class be able to handle a "upgrade to version X" call (which may be a noop if the application is already at the target version).

Reply:
True, and it will make things simpler indeed.

I've refactored the patch and removed apps to upgrade. We will still need to know the current and next OpenStack version for both Application and Analyze, so great part of the code is still valid.

Thanks